ports/stm32/pyb_can.c

```diff
@@ -140,9 +140,57 @@ STATIC void pyb_can_print(const mp_print_t *print, mp_obj_t self_in, mp_print_ki
     }
 }
 
+STATIC void pyb_can_calc_bit_timing(mp_arg_val_t *args) {
+    enum { ARG_mode, ARG_extframe, ARG_prescaler, ARG_sjw, ARG_bs1, ARG_bs2, ARG_auto_restart, ARG_baudrate, ARG_sampling_point};
+
+    uint32_t can_kern_clk = 0;
+    uint32_t baudrate = args[ARG_baudrate].u_int;
+    uint32_t sampoint = args[ARG_sampling_point].u_int;
+
+    // Find CAN kernel clock
+    #if defined(STM32H7)
+    switch (__HAL_RCC_GET_FDCAN_SOURCE()) {
+        case RCC_FDCANCLKSOURCE_HSE:
+            can_kern_clk = HSE_VALUE;
+            break;
+        case RCC_FDCANCLKSOURCE_PLL: {
+            PLL1_ClocksTypeDef pll1_clocks;
+            HAL_RCCEx_GetPLL1ClockFreq(&pll1_clocks);
+            can_kern_clk = pll1_clocks.PLL1_Q_Frequency;
+            break;
+        }
+        case RCC_FDCANCLKSOURCE_PLL2: {
+            PLL2_ClocksTypeDef pll2_clocks;
+            HAL_RCCEx_GetPLL2ClockFreq(&pll2_clocks);
+            can_kern_clk = pll2_clocks.PLL2_Q_Frequency;
+            break;
+        }
+    }
+    #else // F4 and F7 and assume other MCUs too.
+    // CAN1/CAN2/CAN3 on APB1 use GetPCLK1Freq, alternatively use the following:
+    // can_kern_clk = ((HSE_VALUE / osc_config.PLL.PLLM ) * osc_config.PLL.PLLN) /
+    //  (osc_config.PLL.PLLQ * clk_init.AHBCLKDivider * clk_init.APB1CLKDivider);
+    can_kern_clk = HAL_RCC_GetPCLK1Freq();
+    #endif
+    for (int brp = 1; brp < 512; brp++) {// These values work on all MCUs
+        for (int bs1 = 1; bs1 < 16; bs1++) {
+            for (int bs2 = 1; bs2 < 8; bs2++) {
```

Comment:
why not range up to 16?

Reply:
For the ranges on the F4, F7 bs1 and bs2 max are the same (16, 8) and prescaler max is 1024... on the H7 they go up to 256 and 128 and prescaler max 512. So 512, 16 and 8 work on all MCUs, I found out it either converges real quick or doesn't. I can make the ranges go up to the max on each MCU your choice.

Reply:
Actually I think the higher ranges may be for CAN FD mode, the driver doesn't support that.

Reply:
Ok, thanks for the explanation, I think it's good the way it is now (can improve later if needed).